File: 24th_oct/ALGORITHMS/q1/solution.py
#TC : O(n)
#SC : O(1)
# for loop method 
def factorial(x)-> int:
    result = 1      # use 1 because anything multiplied by 0 is 0
    for i in range(x,0,-1):
        result = result*i
       # print("Debug",i)
    return result

def fact_recursive(x)-> int:
    if x==1:
        return 1
    return x*fact_recursive(x-1)

result = fact_recursive(5)
print(result)

File: 24th_oct/ALGORITHMS/q1/solution.cpp
#include<iostream>
using namespace std;

/*
3 
3*2*1
5
5*4*3*2*1

stop at 1, then return
until the next condition: n(n-1)
*/

long long int factorial(int x){
    long long int result;
    result = 1;
    for (int i = x; i>=1; i--){
        result = result*i;
        cout << "Debug" << i << " " << result << endl;
    }
    return result;

} 

// TC: O(n)
int fact_recursive(int x){
    // base case: x = 1 
    if (x==1){
        return 1;
    }
    // operation step: x*fact_recursive(x-1)
    return x*fact_recursive(x-1);

} 

int main(){
    long long int result;
    result = fact_recursive(5);
    cout << result << endl;
    return 0;
}